### PR TITLE
Archive generated pipeline upon request

### PR DESCRIPTION
**Pull request by mdeitrick**
# PR Details

This PR is being submitted to introduce a new function that gives developers the ability to archive a generated pipeline when `jte.archive_generated_pipeline` is set to true.  If the archival of the generated pipeline fails, the fallback method is to log the generated pipeline to the console.

The justification for this PR is to provide developers a stop-gap in the event they need to release an emergency fix to production and JTE fails post-initialization, preventing them from moving forward with the aforementioned fix.  Developers can then apply the generated pipeline to a newly created job, modify the logic to resolve issues observed with the generated JTE pipeline, and get their changes over the finish line.

## Description

A new property - archive_generated_pipeline - will be added to `PipelineDecorator.JteWrapper` and set to false by default. If developers wish to archive the pipeline generated by JTE, then they can set it to true. The `TemplateFlowDefinition` will need to be adjusted to delegate the archiving process to `hudson.tasks.ArtifactArchiver` in conjunction with `hudson.FilePath` to interact with the file system.

## How Has This Been Tested

These changes will be tested by adjusting the `AdHocTemplateFlowDefinition` to incorporate the verification of the archiving process. At least one pipeline will need to be created that utilizes all of JTE's pipeline and library constructors (i.e. hooks) and validate that the archived Jenkinsfile actually contains the pipeline injected with all libraries and respective hooks.

The testing environment includes running all the Gradle tests in a Docker container, as well as manual end-to-end testing to ensure UX is acceptable.

## Types of Changes

- [ ] Added Unit Testing
- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] I have added the appropriate label for this PR
- [ ] If necessary, I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by steven-terrana** [comment]
it's an interesting idea! Thanks for digging into the code. 

I haven't had a chance to deep dive into the code changes just yet but after reading through the rationale i have a few thoughts I wanted to get your perspective on. 

## Declarative Syntax's Resume From Stage

Does Declarative Syntax's [Restart From Stage](https://www.jenkins.io/doc/book/pipeline/running-pipelines/#restart-from-a-stage) functionality help implement this use case if pipeline templates were written in declarative? 

My understanding is that this would resume from the failed stage, which might not work for your use case if you're trying to skip failing stages.

## Library Design Pattern

I'm not sure if you're working in an environment where pipeline governance is a big deal. I know some places would take issue with bypassing failing security scans to deploy directly to production.  Though i suppose if you're able to create a new Jenkins Job from the archived pipeline and modify it - you can already do this by hand anyway. 

regardless - before integrating this into the pipeline i might consider implementing some custom logic into your Pipeline Configuration and Pipeline Libraries. 

Something like:

1. a keyword that serves as a feature flag that enable/disable the ability to skip steps
2. a `skip_steps{}` block in the Pipeline Configuration 
3. some conditional logic that wraps each step to check if the step name is in skip_steps. 

**Pipeline Template**
```groovy
unit_test()
static_code_analysis()
build()
deploy_to prod
```

**Pipeline Configuration**
```groovy
libraries{
  a
  b
}
application_environments{
  prod
}
// custom stuff:
allow_skip_steps = true 
skip_steps = [ "unit_test", "static_code_analysis" ]
```

**Step Logic**
```groovy
void call(){
  /*
    check if:
    1. this step (stepContext.name) is in the skip_steps array
    2. make sure allow_skip_steps is set
  */  
  if(stepContext.name in pipelineConfig.skip_steps && pipelineConfig.allow_skip_steps){
    println "Skipping due to skip_steps configuration in the Pipeline Configuration" 
    return
  }
  /*
    rest of step logic 
  */
}
```

## TL;DR

I'm open the idea, assuming the code's functional and there are unit tests for it. The "blast radius" of the change seems pretty small so far. But before we move forward i'm curious if you feel like either of those options above allows you to achieve the same thing with perhaps the same (or less) effort over having to grab an archived pipeline, modify it, and create a new job out of that new pipeline. 

thoughts?

**Comment by mdeitrick** [comment]
@steven-terrana I've been meaning to follow-up on this thread for a while, but our team decided to use the strategy you mentioned in your post (regarding the addition of a skip flag for skippable libraries).  I think archiving the generated pipeline opens Pandora's box and, frankly, is overengineering a solution for something that is straightforward.  Again, thank you for the constructive feedback!
